src/safeposix/net.rs

```diff
@@ -27,51 +51,124 @@ pub fn mux_port(addr: interface::GenIpaddr, port: u16, domain: i32, istcp: bool)
+            if self.used_port_set.contains_key(tup) {return true;}
+            let mut tupclone = (*tup).clone();
+            match tupclone.2 {
+                PortType::IPv4UDP | PortType::IPv4TCP => {
+                    tupclone.0 = interface::GenIpaddr::V4(interface::V4Addr::default());
+                }
+                PortType::IPv6UDP | PortType::IPv6TCP => {
+                    tupclone.0 = interface::GenIpaddr::V6(interface::V6Addr::default());
+                }
             }
+            self.used_port_set.contains_key(&tupclone)
         }
-        self.used_port_set.contains(&tupclone)
     }
-    pub fn _get_available_udp_port(&mut self, addr: interface::GenIpaddr, domain: i32) -> Result<u16, i32> {
+    pub fn _get_available_udp_port(&mut self, addr: interface::GenIpaddr, domain: i32, rebindability: bool) -> Result<u16, i32> {
+        if !NET_DEVICES_LIST.contains(&addr) {
+            return Err(syscall_error(Errno::EADDRNOTAVAIL, "bind", "Specified network device is not set up for lind or does not exist!"));
+        }
         let mut porttuple = mux_port(addr, 0, domain, UDPPORT);
-        for port in (EPHEMERAL_PORT_RANGE_START ..= EPHEMERAL_PORT_RANGE_END).rev().map(|x| x.to_be()) {
-            porttuple.1 = port;
-            if !self.port_in_use(&porttuple) {
-                self.used_port_set.insert(porttuple);
-                return Ok(port);
+
+        //start from the starting location we specified in a previous attempt to get an ephemeral port
+        let next_ephemeral = if domain == AF_INET {self.next_ephemeral_port_udpv4} else if domain == AF_INET6 {self.next_ephemeral_port_udpv6} else {unreachable!()};
+        for range in [(EPHEMERAL_PORT_RANGE_START ..= next_ephemeral), (next_ephemeral + 1 ..= EPHEMERAL_PORT_RANGE_END)] {
+            for ne_port in range.rev() {
+                let port = ne_port.to_be(); //ports are stored in network endian order
+                porttuple.1 = port;
+
+                //if we think we can bind to this port
+                if !self.port_in_use(&porttuple) {
+                    self.used_port_set.insert(porttuple, if rebindability {1} else {0}); //rebindability of 0 means not rebindable, 1 means it's rebindable and there's 1 bound to it
+
+                    if ne_port == EPHEMERAL_PORT_RANGE_START {
+                        if domain == AF_INET {
+                            self.next_ephemeral_port_udpv4 = EPHEMERAL_PORT_RANGE_END;
+                        } else if domain == AF_INET6 {
+                            self.next_ephemeral_port_udpv6 = EPHEMERAL_PORT_RANGE_END;
+                        } else {unreachable!()};
+                    } else {
+                        if domain == AF_INET {
+                            self.next_ephemeral_port_udpv4 = ne_port - 1;
+                        } else if domain == AF_INET6 {
+                            self.next_ephemeral_port_udpv6 = ne_port - 1;
+                        } else {unreachable!()};
```

Comment:
This is just a small thing, but these unreachable code lines feel unnecessary -- we are just getting the info straight from the Muxport function and the Muxport function has a panic option for unknown inputs. I guess it is safer to not assume that the data is good, but this feels superfluous.

Reply:
I think it's helpful in case we want to implement unix domain sockets later so we can pinpoint failures